kolibri/core/content/utils/content_request.py

```diff
@@ -0,0 +1,525 @@
+            download_request.contentnode_id
+        )
+    )
+    # mark request as processing
+    download_request.status = ContentRequestStatus.InProgress
+    download_request.save()
+
+    try:
+        # by this point we should have a ContentNode
+        channel_id = ContentNode.objects.get(
+            pk=download_request.contentnode_id
+        ).channel_id
+        import_manager = RemoteChannelResourceImportManager(
+            channel_id,
+            peer_id=peer.id,
+            baseurl=peer.base_url,
+            node_ids=[download_request.contentnode_id],
+        )
+        import_manager.run()
+
+        download_request.status = ContentRequestStatus.Completed
+        download_request.save()
+    except Exception as e:
+        logger.exception(e)
+        download_request.status = ContentRequestStatus.Failed
+        download_request.save()
+
+
+def process_user_downloads_for_removal(incomplete_downloads):
+    """
+    Simplistic approach to removing user downloads, starting with largest and working down
+
+    :param incomplete_downloads: a ContentDownloadRequest queryset
+    :type incomplete_downloads: django.db.models.QuerySet
+    """
+    user_downloads = (
+        ContentDownloadRequest.objects.filter(
+            reason=ContentRequestReason.UserInitiated,
+            status=ContentRequestStatus.Completed,
+        )
+        .exclude(
+            contentnode_id__in=incomplete_downloads.values_list(
+                "contentnode_id", flat=True
+            )
+        )
+        .annotate(total_size=_total_size_annotation())
+    )
+
+    # TODO: add more sophisticated logic for choosing which user downloads to remove, like based
+    # off the user's interaction with the resource, e.g. complete status
+    largest_user_download = user_downloads.order_by("-total_size").first()
+
+    # adding this opposite of the user download request allows us to detect this situation
+    user_download_removal = ContentRemovalRequest(
+        facility_id=largest_user_download.facility_id,
+        source_model=largest_user_download.source_model,
+        source_id=largest_user_download.source_id,
+        reason=ContentRequestReason.SyncInitiated,
+        status=ContentRequestStatus.Pending,
+        contentnode_id=largest_user_download.contentnode_id,
+    )
+    # this removal request will be processed on the next loop
+    user_download_removal.save()
+    logger.info(
+        "Added removal request for user download of {}".format(
+            largest_user_download.contentnode_id
+        )
+    )
+
+
+def process_content_removal_requests(queryset):
+    """
+    Garbage collects content requests marked for removal (removed_at is not null)
+
+    :param queryset: a ContentRemovalRequest queryset
+    :type queryset: django.db.models.QuerySet
+    """
+    # exclude admin imported nodes
+    removable_nodes = ContentNode.objects.filter(
+        Q(admin_imported__isnull=True)
```

Comment:
This should only be happening in develop environments, as the upgrade handler should prevent this ever actually being null - just in case this has performance implications.